rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/attachment/Sw360AttachmentService.java

```diff
@@ -68,9 +97,92 @@ public AttachmentInfo getAttachmentByIdForUser(String id, User sw360User) throws
+            response.setContentType(contentType);
+            response.setHeader("Content-Disposition", String.format("attachment; filename=\"%s\"", filename));
+            FileCopyUtils.copy(attachmentStream, response.getOutputStream());
+        } catch (TException|IOException e) {
+            log.error(e.getMessage());
+        }
+    }
+
+    public <T> InputStream getStreamToAttachments(Set<AttachmentContent> attachments, User sw360User, T context) throws IOException, TException {
+        return new AttachmentFrontendUtils().getStreamToServeAFile(attachments, sw360User, context);
+    }
+
+    public Attachment uploadAttachment(MultipartFile file, Attachment newAttachment, User sw360User) throws IOException, TException {
+        String fileName = file.getOriginalFilename(); // TODO: shouldn't the fileName be taken from newAttachment?
+        String contentType = file.getContentType();
+        final AttachmentContent attachmentContent = makeAttachmentContent(fileName, contentType);
+
+        final AttachmentConnector attachmentConnector = getConnector();
+        Attachment attachment = new AttachmentFrontendUtils().uploadAttachmentContent(attachmentContent, file.getInputStream(), sw360User);
+        attachment.setSha1(attachmentConnector.getSha1FromAttachmentContentId(attachmentContent.getId()));
+
+        AttachmentType attachmentType = newAttachment.getAttachmentType();
+        if (attachmentType != null) {
+            attachment.setAttachmentType(attachmentType);
+        }
+
+        CheckStatus checkStatus = newAttachment.getCheckStatus();
+        if (checkStatus != null) {
+            attachment.setCheckStatus(checkStatus);
+        }
+
+        return attachment;
+    }
+
+    private AttachmentContent makeAttachmentContent(String filename, String contentType) {
+        AttachmentContent attachment = new AttachmentContent()
+                .setContentType(contentType)
+                .setFilename(filename)
+                .setOnlyRemote(false);
+        return makeAttachmentContent(attachment);
+    }
+
+    private AttachmentContent makeAttachmentContent(AttachmentContent content) {
+        try {
+            return new AttachmentFrontendUtils().makeAttachmentContent(content);
+        } catch (TException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public AttachmentContent getAttachmentContent(String id) {
+        try {
+            return new AttachmentFrontendUtils().getAttachmentContent(id);
+        } catch (TException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
     private ComponentService.Iface getThriftComponentClient() throws TTransportException {
         THttpClient thriftClient = new THttpClient(thriftServerUrl + "/components/thrift");
         TProtocol protocol = new TCompactProtocol(thriftClient);
         return new ComponentService.Client(protocol);
     }
+
+    private AttachmentConnector getConnector() throws TException {
```

Comment:
This method can also be defined as synchronized.

Reply:
No, it shouldn't be synchronized. Why would you want to prevent parallel reads?